### PR TITLE
Claim full support for Java 17.

### LINKED ISSUE #12838: Java 17 compatibility
Java 17 is the current LTS release. We'd like to officially support this.

### PR DESCRIPTION
**Pull request by gianm**
~No production code has changed, except the startup scripts.~

Some production Java code has changed:

1) Add opens and exports in ForkingTaskRunner, rather than relying on
   `bin/run-java`. I expect this would be more robust.

2) Minor changes to fix or suppress issues noticed by the errorprone upgrade.

Script, doc, test changes:

1) Allow Java 17 without DRUID_SKIP_JAVA_CHECK.

2) Include the full list of opens and exports on both Java 11 and 17.

3) Document that Java 17 is both supported and preferred.

4) Switch some tests from Java 11 to 17 to get better coverage on the
   preferred version.

5) Update errorprone to a version that supports Java 17.

Fixes #12838.

### CODE REVIEW

**Review by clintropolis — APPROVED** [review]
sgtm, :+1: (after CI, didn't look into why it is failing)

I was using java 17 with no issue for quite some time in most of my local development until fairly recently where I have switched to java 20 (which has some dependency problems for some tests such as the equalsverifier stuff which maybe needs to be updated to be cool).

**Review by maytasm — APPROVED** [review]

File: docs/operations/java.md [hunk 1]
```diff
@@ -85,29 +80,22 @@ these JVMs is given by `druid.indexer.runner.javaOptsArray` or `druid.indexer.ru
...
+druid.indexer.runner.javaOptsArray=["-server","-Xms1g","-Xmx1g","-XX:MaxDirectMemorySize=1g","-Duser.timezone=UTC","-Dfile.encoding=UTF-8","-XX:+ExitOnOutOfMemoryError","-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager","--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED","--add-exports=java.base/jdk.internal.ref=ALL-UNNAMED","--add-opens=java.base/java.nio=ALL-UNNAMED","--add-opens=java.base/sun.nio.ch=ALL-UNNAMED","--add-opens=java.base/jdk.internal.ref=ALL-UNNAMED","--add-opens=java.base/java.io=ALL-UNNAMED","--add-opens=java.base/java.lang=ALL-UNNAMED","--add-opens=jdk.management/com.sun.management.internal=ALL-UNNAMED"]
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/operations/java.md:83]: `bin/run-java` does that, so it does happen if you set `druid.indexer.runner.javaCommand=bin/run-java` (which the default configs do). Having the `ForkingTaskRunner` do it also seems legit though.

**Comment by gianm** [comment]
> I was using java 17 with no issue for quite some time in most of my local development until fairly recently where I have switched to java 20 (which has some dependency problems for some tests such as the equalsverifier stuff which maybe needs to be updated to be cool).

Well, Java 20 will be next 🙂

**Comment by gianm** [comment]
Raised https://github.com/apache/druid/issues/14502 for Java 21.

**Comment by maytasm** [comment]
@gianm Do we need to update java.version, maven.compiler.source and maven.compiler.target to Java 17?

**Pull request merged**